File: 02_NLP/JS/03 cosine.js
const fs = require('fs')
const talismanTokenaizer = require('talisman/tokenizers/words/treebank'); // treebank Для англ.языка - разбивает в т.ч. сокращения
// const talismanTokenaizer = require('talisman/tokenizers/words/naive') // naive без знаков припинания

const cosine = require('talisman/metrics/cosine')

// https://github.com/akumidv/math-phyton/blob/master/7-11.programmir.ipynb

const catsFile = fs.readFileSync('./sentences.txt').toString().toLowerCase()
const catsText = catsFile.split('\n')
console.log(`\nПРЕДЛОЖЕНИЯ. ВСЕГО ${catsText.length}`)
console.log(catsText.slice(0,5))


const catSentenceTokens = catsText
  .map(sentence => talismanTokenaizer(sentence).filter(item => /^[a-z]/.test(item)))
console.log(`\nТОКЕНЫ ПРИЛОЖЕНИЙ`)
console.log(catSentenceTokens.slice(0,1))


const catTextTokens = talismanTokenaizer(catsFile)
  .filter((word, i, arr) => /^[a-z]/.test(word) && arr.indexOf(word) === i)
console.log(`\nВСЕ ТОКЕНЫ. ВСЕГО ${catTextTokens.length}`)
console.log(catTextTokens.slice(0,10))


const wordDic = {}
// catTextTokens.forEach((item, index) => wordDic[item] = index)
let index = 0
catSentenceTokens.forEach(sentence => {
  sentence.forEach((word) => typeof wordDic[word] === 'undefined' ?
    wordDic[word] = index++ : wordDic[word])
})
console.log(`\nСЛОВАРЬ. ВСЕГО ${Object.keys(wordDic).length}`)
index = 0
for(let key in wordDic) {
  console.log(`  ${key}:${wordDic[key]}`)
  if (index++ > 10) break
}

const wordInclude = catSentenceTokens.map(sentence => {
  const sentenceRow = Object.keys(wordDic).map(() => 0)
  sentence.forEach(word => sentenceRow[wordDic[word]]++)
  return sentenceRow
})
console.log(`\nВЕКТОР ПЕРВОГО ПРЕДЛОЖЕНИЯ. ДЛИНА ВЕКТОРА ${wordInclude[0].length}`)
console.log(wordInclude[0])

const dist = wordInclude.map(sentence => (1 - cosine(wordInclude[0], sentence)))
console.log('\nДИСТАНЦИИ МЕЖДУ ПРЕДЛОЖЕНИЯМИ')
console.log(dist.slice(0,5))

const minVal = dist.concat().sort((a,b) => a-b)
console.log('\nИСКОМАЯ СТРОКА:', 0, 'ДИСТАНЦИЯ:', 0, '\n', catsText[0])
console.log('\nБЛИЖАЙШАЙШАЯ СТРОКА:', dist.indexOf(minVal[1]), 'ДИСТАНЦИЯ:', minVal[1], '\n', catsText[dist.indexOf(minVal[1])])
console.log('\nБЛИЖАЙШАЙШАЯ СТРОКА2:', dist.indexOf(minVal[2]), 'ДИСТАНЦИЯ:', minVal[2], '\n', catsText[dist.indexOf(minVal[2])])

File: 02_NLP/Python/03 cosine.py
import re
import numpy as np
import scipy.spatial

# Сравнение предложений
catsFile = open('./sentences.txt', 'r')

catsText = catsFile.read().lower().split('\n')
print('ПРЕДЛОЖЕНИЯ. ВСЕГО', len(catsText))
print('\n'.join(catsText[0:5]))

catsToken = []  # Строки с разделением на слова
wordDic = dict()  # Словарь частот вхождения слов

for sent in catsText:
    tmpStr = re.split('[^a-z]', sent)  # Разбиваем на слова, без символов
    strKey = []
    for key in tmpStr:  # Очищаем массив от пустых элементов
        if key:
            strKey.append(key)
            wordDic.setdefault(key)  # Формируем словарь
    catsToken.append(strKey)
print('\nТОКЕНЫ ПРИЛОЖЕНИЙ')
print(catsToken[0:5])

index = 0
for key in wordDic:  # Проставляем индекс у каждого слова
    wordDic[key] = index
    index += 1
print('\nСЛОВАРЬ. ВСЕГО', len(wordDic))
print(wordDic)

wordInclude = np.zeros((len(catsText),len(wordDic)))  # Массив размерностью строк и числа встреченных слов
sentInd = 0
for sentence in catsToken:
    for word in sentence:
        wordInclude[sentInd][wordDic[word]] += 1
    sentInd += 1
print('ВЕКТОР ПЕРВОГО ПРЕДЛОЖЕНИЯ. ДЛИНА ВЕКТОРА', len(wordInclude[0]))
print(wordInclude[0])

dists = []
for row in wordInclude:
    dists.append(scipy.spatial.distance.cosine(wordInclude[0,:], row))
print('\nДИСТАНЦИИ МЕЖДУ ПРЕДЛОЖЕНИЯМИ')
print(dists[0:5])

i = 0
min1 = 1
min2 = 1
i1 = 0
i2 = 0
for value in dists:  # Ищем минимальных два индекса, ближайших и их значнеия
    if (value != 0.) & (value < min1):
        min2 = min1
        min1 = value
        i2 = i1
        i1 = i
    i += 1
print('\nИСКОМАЯ СТРОКА:', 0, 'ДИСТАНЦИЯ:', 0, '\n', catsText[0])
print('\nБЛИЖАЙШАЙШАЯ СТРОКА:', i1, 'ДИСТАНЦИЯ:', min1, '\n', catsText[i1])
print('\nБЛИЖАЙШАЙШАЯ СТРОКА2:', i2, 'ДИСТАНЦИЯ:', min2, '\n', catsText[i2])